Feat/add miner deposit v3

tediou5 opened the pull request:
Fip: [FIP-0077](https://github.com/filecoin-project/FIPs/pull/796)
Fip update: [FIP-0077 update](https://github.com/filecoin-project/FIPs/pull/963)
Discussions: [discussions](https://github.com/filecoin-project/FIPs/discussions/780)

Also, should we allow users to use the deposit as an initial pledge? I think i can do this.

anorth reviewed (commented):
I'm reviewing this as a high level implementation proposal rather than in detail yet.

There are two decisions here that cause a much larger impact on code and complexity that I think are necessary.
- New state for the deposit. We already have a vesting table capable of representing the idea of funds locked until some epoch. Adding new state representing a new type of locked funds means all the calculations about balances need to take that into account. The chance of introducing an error now or later is very high. It requires a state migration. I can't see any advantage of the separate representation.
- A new method for the deposit. The deposit is required at miner creation. It would be simpler to just pass the funds in the call to the miner constructor (which already happens) and compute the lock in that constructor. The duplication of miner logic code in the power actor will not be acceptable (though we could extract it somewhere to share). The power actor is already calling the miner actor once, there's no point calling a second time. The only parameter is the lock amount, which can be computed at the miner, so we don't need to adjust the miner constructor parameters. The constructor can fail if the balance is insufficient for the lock.

anorth commented:
> should we allow users to use the deposit as an initial pledge

I don't think this is worth the complexity it will involve to implement it.  I suggest not doing that for now.

tediou5 commented:
@anorth 

> A new method for the deposit. The deposit is required at miner creation. It would be simpler to just pass the funds in the call to the miner constructor (which already happens) and compute the lock in that constructor. The duplication of miner logic code in the power actor will not be acceptable (though we could extract it somewhere to share). The power actor is already calling the miner actor once, there's no point calling a second time. The only parameter is the lock amount, which can be computed at the miner, so we don't need to adjust the miner constructor parameters. The constructor can fail if the balance is insufficient for the lock.

Yeah, I'm just want an early exit here. I will fix this.

------

> New state for the deposit. We already have a vesting table capable of representing the idea of funds locked until some epoch. Adding new state representing a new type of locked funds means all the calculations about balances need to take that into account. The chance of introducing an error now or later is very high. It requires a state migration. I can't see any advantage of the separate representation.

Yes, adding a state makes it a lot more complicated. But its very easy and necessary to use the deposit as an initial pledge, if we have an deposit state. 

Maybe we could give it a try ? Or if you're very adamant here, I can remove the code.

anorth commented:
> Yes, adding a state makes it a lot more complicated. But its very easy and necessary to use the deposit as an initial pledge, if we have an deposit state.

My opinion is that it is not worthwhile to have that feature. I'm open to other opinions, more information etc though. My default stance is that any additional complexity has to bring a lot of benefit.

rjan90 commented:
Hey! 👋

As part of our cleanup to kick off the year, I'm reviewing all open non-draft pull requests. Could you please do one of the following for your PR?

**1. Close it:** If it's no longer needed.
**2. Mark as Draft:** If it needs more work, and add the next steps. 
**3. Ready for Review:** If it's good to go, let me know, and I'll assign a reviewer.

If there's no response in a week, I'll assume it's option 1 and close the PR. If you have any questions, just let me know.

Thanks for your help in keeping things organized, and I appreciate your contributions!

tediou5 commented:
> Hey! 👋
> 
> As part of our cleanup to kick off the year, I'm reviewing all open non-draft pull requests. Could you please do one of the following for your PR?
> 
> **1. Close it:** If it's no longer needed. **2. Mark as Draft:** If it needs more work, and add the next steps. **3. Ready for Review:** If it's good to go, let me know, and I'll assign a reviewer.
> 
> If there's no response in a week, I'll assume it's option 1 and close the PR. If you have any questions, just let me know.
> 
> Thanks for your help in keeping things organized, and I appreciate your contributions!

@rjan90 Thank you for your work. This PR has been cleaned up and re-implemented according to our previous discussions. It is now ready for review.

BigLep commented:
@tediou5 : I see this slipped through the maintainer cracks.  I'm actually not sure how.  Regardless, can you resolve conflicts and we'll get a reviewer assigned?

@rvagg : do you have a reviewer suggestion for this one?

rvagg commented:
Are we in a chicken/egg situation with this and https://github.com/filecoin-project/FIPs/pull/963? @tediou5 are you still keen to pursue attempting to get FIP-0077 accepted? If so, how are you proposing to proceed here? Do you want to get some clarity on the implementation here and then go reopen https://github.com/filecoin-project/FIPs/pull/963, or do you think that you've got enough feedback from the existing work with Alex so far to go back and do that?
I'm not yet up to speed on this FIP, other than remembering the CreateMiner spam incident, but can get up to speed if you need help getting the implementation further if that's what's needed to get the FIP update reopened so core-devs can consider it.

tediou5 commented:
> Are we in a chicken/egg situation with this and [filecoin-project/FIPs#963](https://github.com/filecoin-project/FIPs/pull/963)? @tediou5 are you still keen to pursue attempting to get FIP-0077 accepted?

Yes, I’ll communicate with @ipfsforcezuofu and reopen [filecoin-project/FIPs#963](https://github.com/filecoin-project/FIPs/pull/963).

tediou5 commented:
@BigLep Conflicts and test cases have been fixed.

tediou5 commented:
Rebase for fix conflicts and test cases.

rvagg commented on actors/miner/src/lib.rs:
```diff
@@ -5627,6 +5644,40 @@ fn activate_new_sector_infos(
+        ramp_duration_epochs: pwr.ramp_duration_epochs,
+    };
+
+    let sector_number = MINIMUM_CONSENSUS_POWER / sector_size as i64;
+    let power = qa_power_for_weight(sector_size, MIN_SECTOR_EXPIRATION, &BigInt::zero());
+    let sector_initial_pledge = initial_pledge_for_power(
+        &power,
+        &pledge_inputs.network_baseline,
+        &pledge_inputs.epoch_reward,
+        &pledge_inputs.network_qap,
+        &pledge_inputs.circulating_supply,
+        pledge_inputs.epochs_since_ramp_start,
+        pledge_inputs.ramp_duration_epochs,
+    );
+    Ok(sector_initial_pledge * sector_number)
```
I'm not following the way `sector_number` is used here, maybe we're juggling the "weight" nonsense in doing this? Wouldn't a more direct way to do this just to be to calculate it something like this:

```suggestion
    let power = BigInt::from(MINIMUM_CONSENSUS_POWER);
    let minimum_consensus_pledge = initial_pledge_for_power(
        &power,
        &pledge_inputs.network_baseline,
        &pledge_inputs.epoch_reward,
        &pledge_inputs.network_qap,
        &pledge_inputs.circulating_supply,
        pledge_inputs.epochs_since_ramp_start,
        pledge_inputs.ramp_duration_epochs,
    );
    Ok(minimum_consensus_pledge)
```

?

rvagg commented on actors/miner/tests/miner_actor_test_construction.rs:
```diff
@@ -100,10 +145,21 @@ fn simple_construction() {
 
     assert_eq!(TokenAmount::zero(), state.pre_commit_deposits);
-    assert_eq!(TokenAmount::zero(), state.locked_funds);
+    assert_eq!(TokenAmount::from_atto(633318697598976000u64), state.locked_funds);
```
ditto here, it would be good to involve some constants or named variables to at least describe the starting point and the calculation used to arive here

rvagg commented on actors/miner/src/lib.rs:
```diff
@@ -5627,6 +5644,40 @@ fn activate_new_sector_infos(
+/// See FIP-0077, https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0077.md
+pub fn calculate_create_miner_deposit(
+    rt: &impl Runtime,
+    network_qap: FilterEstimate,
```
don't need this, you get it from `pwr` below

rvagg commented on actors/miner/src/lib.rs:
```diff
@@ -5627,6 +5644,40 @@ fn activate_new_sector_infos(
+    let pwr = request_current_total_power(rt)?;
+    let circulating_supply = rt.total_fil_circ_supply();
+    let pledge_inputs = NetworkPledgeInputs {
+        network_qap,
```
```suggestion
        pwr.quality_adj_power_smoothed,
```

rvagg commented on actors/power/src/ext.rs:
```diff
@@ -46,6 +46,7 @@ pub mod miner {
         #[serde(with = "strict_bytes")]
         pub peer_id: Vec<u8>,
         pub multi_addresses: Vec<BytesDe>,
+        pub network_qap: FilterEstimate,
```
shouldn't be needed, see comment in miner actor

BigLep commented:
@tediou5 : will you be able to incorporate the corresponding feedback?

tediou5 commented:
> will you be able to incorporate the corresponding feedback?

@BigLep Of course. I'm currently addressing the feedback from the previous review and making adjustments according to the latest FIP.

BigLep commented:
Thanks for reviewing @ZenGround0 .

@tediou5 : will you be able to incorporate comments by end of day 2025-08-29?

tediou5 commented:
@BigLep @ZenGround0 68e18aa, Done.

BigLep commented:
Followup Lotus work is being tracked in https://github.com/filecoin-project/lotus/issues/13287
